hw/dv/dpi/gpiodpi/gpiodpi.c

```diff
@@ -186,53 +198,84 @@ static uint32_t parse_dec(char **text) {
+                                     svBitVecVal *gpio_pull_sel) {
   struct gpiodpi_ctx *ctx = (struct gpiodpi_ctx *)ctx_void;
   assert(ctx);
 
-  char gpio_str[32 + 2];
-  ssize_t read_len = read(ctx->host_to_dev_fifo, gpio_str, 32 + 1);
-  if (read_len < 0) {
-    return ctx->driven_pin_values;
-  }
-  gpio_str[read_len] = '\0';
-
-  char *gpio_text = gpio_str;
-  for (; *gpio_text != '\0'; ++gpio_text) {
-    switch (*gpio_text) {
-      case '\n':
-      case '\r':
-      case '\0':
-        goto parse_loop_end;
-      case 'l':
-      case 'L': {
-        ++gpio_text;
-        int idx = parse_dec(&gpio_text);
-        if (!GET_BIT(gpio_oe[0], idx)) {
-          fprintf(stderr,
-                  "GPIO: Host tried to pull disabled pin low: pin %2d\n", idx);
-        }
-        CLR_BIT(ctx->driven_pin_values, idx);
-        break;
-      }
-      case 'h':
-      case 'H': {
-        ++gpio_text;
-        int idx = parse_dec(&gpio_text);
-        if (!GET_BIT(gpio_oe[0], idx)) {
-          fprintf(stderr,
-                  "GPIO: Host tried to pull disabled pin high: pin %2d\n", idx);
+  if (ctx->counter % TICKS_PER_SYSCALL == 0) {
+    char gpio_str[32 + 2];
+    ssize_t read_len = read(ctx->host_to_dev_fifo, gpio_str, 32 + 1);
+    if (read_len > 0) {
+      gpio_str[read_len] = '\0';
+
+      bool weak = false;
+      char *gpio_text = gpio_str;
+      for (; *gpio_text != '\0'; ++gpio_text) {
+        switch (*gpio_text) {
+          case '\n':
+          case '\r':
+          case '\0':
+            goto parse_loop_end;
+          case 'w':
+          case 'W': {
+            weak = true;
+            break;
+          }
+          case 'l':
+          case 'L': {
+            ++gpio_text;
+            int idx = parse_dec(&gpio_text);
+            if (!GET_BIT(gpio_oe[0], idx)) {
+              fprintf(stderr,
+                      "GPIO: Host tried to pull disabled pin low: pin %2d\n",
+                      idx);
+            }
+            CLR_BIT(ctx->driven_pin_values, idx);
+            set_bit_val(&ctx->weak_pins, idx, weak);
+            weak = false;
+            break;
+          }
+          case 'h':
+          case 'H': {
+            ++gpio_text;
+            int idx = parse_dec(&gpio_text);
+            if (!GET_BIT(gpio_oe[0], idx)) {
+              fprintf(stderr,
+                      "GPIO: Host tried to pull disabled pin high: pin %2d\n",
+                      idx);
+            }
+            SET_BIT(ctx->driven_pin_values, idx);
+            set_bit_val(&ctx->weak_pins, idx, weak);
+            weak = false;
+            break;
+          }
+          default:
+            break;
         }
-        SET_BIT(ctx->driven_pin_values, idx);
-        break;
       }
-      default:
-        break;
     }
   }
 
 parse_loop_end:
-  return ctx->driven_pin_values;
+  ctx->counter += 1;
+  // Candidates for reporting a value other than the driven value are those pins
```

Comment:
could you add this comment block, or reference it in the `verilator.sv` file just so hw folks in the future won't look at it and be confused why it's there?

Reply:
Added a comment to chip_earlgrey_verilator.sv.